pkg/controller/buildrun/runtime_image.go

```diff
@@ -0,0 +1,194 @@
+	v1 "k8s.io/api/core/v1"
+)
+
+const (
+	// runtimeDockerfileTmpl Dockerfile template to be used with runtime-image, it uses Build attributes
+	// directly as template input.
+	runtimeDockerfileTmpl = `FROM {{ .Spec.Output.ImageURL }} as builder
+
+FROM {{ .Spec.Runtime.Base.ImageURL }}
+
+{{- range $k, $v := .Spec.Runtime.Env }}
+ENV {{ $k }}="{{ $v }}"
+{{- end }}
+
+{{- range $k, $v := .Spec.Runtime.Labels }}
+LABEL {{ $k }}="{{ $v }}"
+{{- end }}
+
+{{- range $dir := .Spec.Runtime.Directories }}
+{{- $parts := splitDirectories $dir }}
+COPY --from=builder "{{ index $parts 0 }}" "{{ index $parts 1 }}"
+{{- end }}
+
+{{- if .Spec.Runtime.WorkDir }}
+WORKDIR "{{ .Spec.Runtime.WorkDir }}"
+{{- end }}
+
+{{- if .Spec.Runtime.Entrypoint }}
+ENTRYPOINT [ {{ renderEntrypoint .Spec.Runtime.Entrypoint }} ]
+{{- end -}}
+`
+
+	// workspaceDir common workspace directory, where the source code is located.
+	workspaceDir = "/workspace/source"
+
+	// kanikoStable stable buildah image.
+	kanikoStable = "gcr.io/kaniko-project/executor:v0.23.0"
+
+	// runtimeDockerfile runtime Dockerfile file name.
+	runtimeDockerfile = "Dockerfile.runtime"
+)
+
+// rootUserID root's UID
+var rootUserID = int64(0)
+
+// runtimeDockerfilePath path to runtime Dockerfile on workspace directory.
+var runtimeDockerfilePath = path.Join(workspaceDir, runtimeDockerfile)
+
+// splitDirectories split informed directory by colon, returning a slice with parts. When colon is
+// not present, return the informed directory twice. This method always returns a slice with two
+// entries.
+func splitDirectories(dir string) []string {
+	parts := strings.Split(dir, ":")
+	if len(parts) == 2 {
+		return parts
+	}
+	return []string{dir, dir}
+}
+
+// renderEntrypoint will take a slice of strings and render the notation expected on ENTRYPOINT.
+func renderEntrypoint(e []string) string {
+	entrypoint := []string{}
+	for _, cmd := range e {
+		entrypoint = append(entrypoint, strconv.Quote(cmd))
+	}
+	return strings.Join(entrypoint, ", ")
+}
+
+// renderRuntimeDockerfile render runtime Dockerfile using build instance and pre-defined template.
+func renderRuntimeDockerfile(b *buildv1alpha1.Build) (*bytes.Buffer, error) {
+	tmpl, err := template.New(runtimeDockerfile).
+		Funcs(template.FuncMap{
+			"splitDirectories": splitDirectories,
+			"renderEntrypoint": renderEntrypoint,
+		}).
+		Parse(runtimeDockerfileTmpl)
+	if err != nil {
+		return nil, err
+	}
+
+	dockerfile := new(bytes.Buffer)
+	if err = tmpl.Execute(dockerfile, b); err != nil {
+		return nil, err
+	}
+	return dockerfile, nil
+}
+
+// runtimeDockerfileStep trigger the rendering of Dockerfile.runtime, and use this input as a
+// build-step to create a new file.
+func runtimeDockerfileStep(b *buildv1alpha1.Build) (*buildv1alpha1.BuildStep, error) {
+	dockerfile, err := renderRuntimeDockerfile(b)
+	if err != nil {
+		return nil, err
+	}
+
+	container := v1.Container{
+		Name:  "runtime-dockerfile",
+		Image: b.Spec.BuilderImage.ImageURL,
```

Comment:
I do not think that this should be user-controllable. We should decide for one image (ubi-minimal?) and make it configurable like the kaniko image.

Reply:
I think ultimately the runtime base-image should be in control of the final users. (cc @sbose78)

Reply:
I agree, the runtime base image is your feature and the user specifies it under `spec.runtime.image`. But, this line of code is specifying which container is used to `echo` the Dockerfile-runtime to disk. And the container image to do this should not be configurable in my opinion.

Reply:
It does create a file (`Dockerfile.runtime`) in the `/workspace/source` directory, which can be considered a temporary file. Before the build starts, Tekton populates a new `/workspace/source` directory with data coming from Git repository only.

Reply:
@otaviof I completely get the purpose of this step and I compare it with the `step-create-dir-image` that Tekton runs before even doing the `git clone` as the following step. And `step-create-dir-image` is using the `busybox` image. I as the user defining the build and buildrun (and even the admin defining the cluster build strategy) have no way to define which image is used there (correct me, if that is possible through some means).

Reply:
@SaschaSchwarze0 Right now the [step which materializes the `Dockerfile.runtime`](https://github.com/otaviof/build/blob/ccae990c7422f85b9d8ba205f26a6a68cb896f47/pkg/controller/buildrun/runtime_image.go#L136) does follow the same builder-image than the rest of the strategy. Therefore, I think we can say it's configurable.

Reply:
No, I have to disagree. Your added e2e test specifies `heroku/buildpacks:18` which is incorrect. For the buildpacks cluster build strategies, the builder image is hard-coded in the build strategy since https://github.com/redhat-developer/build/pull/157. A buildpack build since then does not anymore require the builder image to be set in the build. The only build strategy remaining that requires the builder image is source2image. So, my expectation is that the runtime image support works without specifying a builder image. Is that the case? And if yes, which image is taken then?

Reply:
> No, I have to disagree. Your added e2e test specifies `heroku/buildpacks:18` which is incorrect. For the buildpacks cluster build strategies, the builder image is hard-coded in the build strategy since #157. A buildpack build since then does not anymore require the builder image to be set in the build. 

Okay, fair enough.

> The only build strategy remaining that requires the builder image is source2image. So, my expectation is that the runtime image support works without specifying a builder image. Is that the case? And if yes, which image is taken then?

Since all we need from the image is a shell and its builtin `echo`, it's fair to assume most of the container building images will contain those requirements. For the `s2i` image shipped from Red Hat, it does contain the required items.

However, the image used in `s2i` strategy is chosen by the user, and in this situation, this requirement may not be fulfilled in the case of not using images we control. So, by knowing that in the future we will control the builder-images in all strategies, we can quite easily make sure it contains a  `shell` and `echo`.

I think it's a good move to re-use the same image for this step, like for instance by using the local node cache, given the context.

Reply:
I think this discussion goes a little bit too much back and forth. Two ideas:

* we have a slack chat tomorrow (I'm off Thu/Fri his week) to discuss it
* we consolidate our opinions, ideas and concerns specifically to this item somewhere and discuss and decide on next Monday' community call

Reply:
@SaschaSchwarze0 lets bring this during community meeting then. Thanks for the heads up.